rs/types/registry_canister_types/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+use std::fmt;
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum IPv4ConfigError {
+    #[error("Invalid IPv4 address")]
+    InvalidIPv4Address,
+    #[error("Invalid gateway IPv4 address")]
+    InvalidGatewayAddress,
+    #[error("Invalid prefix length")]
+    InvalidPrefixLength,
+    #[error("IP addresses are not in the same subnet")]
+    NotInSameSubnet,
+    #[error("IPv4 address is not a global address")]
+    NotGlobalIPv4Address,
+}
+
+fn is_valid_ipv4_prefix_length(prefix_length: u32) -> bool {
+    prefix_length <= 32
+}
+
+fn is_global_ipv4_address(ipv4_address: &str) -> bool {
+    let ip_address = Ipv4Addr::from_str(ipv4_address).unwrap();
+
+    // IETF RFC 2544
+    let is_benchmarking = u32::from_be_bytes(ip_address.octets()) & (!0 << 17)
+        == u32::from_be_bytes(Ipv4Addr::new(198, 18, 0, 0).octets());
+    // IETF RFC 1112
+    let is_reserved = u32::from_be_bytes(ip_address.octets()) & (!0 << 28)
+        == u32::from_be_bytes(Ipv4Addr::new(240, 0, 0, 0).octets());
+    // IETF RFC 6598
+    let is_shared = u32::from_be_bytes(ip_address.octets()) & (!0 << 22)
+        == u32::from_be_bytes(Ipv4Addr::new(100, 64, 0, 0).octets());
+
+    !is_benchmarking
+        && !ip_address.is_broadcast()
+        && !ip_address.is_documentation()
+        && !ip_address.is_link_local()
+        && !ip_address.is_private()
+        && !ip_address.is_loopback()
+        && !is_reserved
+        && !is_shared
+        && !ip_address.is_unspecified()
+}
+
+// Check that the given IPv4 addresses are all in the same subnet with respect to the subnet mask
+fn are_in_the_same_subnet(ipv4_addresses: Vec<String>, prefix_length: u32) -> bool {
+    let bitmask: u32 = !0 << (32 - prefix_length);
+    let bitmask: [u8; 4] = bitmask.to_be_bytes();
+    let subnet_mask = Ipv4Addr::new(bitmask[0], bitmask[1], bitmask[2], bitmask[3]);
+
+    let subnet_prefixes: HashSet<Vec<u8>> = ipv4_addresses
+        .iter()
+        .map(|x| extract_subnet_bytes(x.as_str(), subnet_mask))
+        .collect();
+    // all IP addresses are in the same subnet, if they share the same subnet bytes
+    subnet_prefixes.len() == 1
+}
+
+// Helper function to extract the subnet bytes of the given IP address
+fn extract_subnet_bytes(ipv4_address: &str, subnet_mask: Ipv4Addr) -> Vec<u8> {
+    Ipv4Addr::from_str(ipv4_address)
+        .unwrap_or_else(|_| panic!("Failed to parse IP address: {}", ipv4_address))
+        .octets()
+        .iter()
+        .zip(subnet_mask.octets().iter())
+        .map(|(&a, &b)| a & b)
+        .collect::<Vec<u8>>()
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct IPv4Config {
+    ip_addr: String,
+    gateway_ip_addr: String,
+    prefix_length: u32,
+}
+
+impl IPv4Config {
+    pub fn new(
+        ip_addr: String,
+        gateway_ip_addr: String,
+        prefix_length: u32,
+    ) -> Result<Self, IPv4ConfigError> {
```

Comment:
I think this could make it easier and possibly more efficient to pass arguments, since we wouldn't have to convert to String:
```suggestion
    pub fn new<S: AsRef<str>>(
        ip_addr: S,
        gateway_ip_addr: S,
        prefix_length: u32,
    ) -> Result<Self, IPv4ConfigError> {
```

Reply:
We will need to clone anyways because the data fields are Strings.